include/boost/geometry/algorithms/detail/closest_points/point_to_geometry.hpp

```diff
@@ -0,0 +1,454 @@
+
+
+namespace boost { namespace geometry
+{
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail { namespace closest_points
+{
+
+struct point_to_point
+{
+    template <typename P1, typename P2, typename Segment, typename Strategies>
+    static inline void apply(P1 const& p1, P2 const& p2,  
+                             Segment& shortest_seg, Strategies const&)
+    {
+        set_segment_from_points::apply(p1, p2, shortest_seg);
+    }
+};
+
+struct point_to_segment
+{
+    template <typename Point, typename Segment, typename OutputSegment, typename Strategies>
+    static inline void apply(Point const& point, Segment const& segment,
+                             OutputSegment& shortest_seg, Strategies const& strategies)
+    {
+        typename point_type<Segment>::type p[2];
+        geometry::detail::assign_point_from_index<0>(segment, p[0]);
+        geometry::detail::assign_point_from_index<1>(segment, p[1]);
+
+        boost::ignore_unused(strategies);
+
+        auto closest_point = strategies.closest_points(point, segment)
+            .apply(point, p[0], p[1]);
+
+        set_segment_from_points::apply(point, closest_point, shortest_seg);    
+    }
+};
+
+/*
+struct point_to_box
+{
+    template<typename Point, typename Box, typename Strategies>
+static inline auto apply(Point const& point, Box const& box,
+                             Strategies const& strategies)
+    {
+        boost::ignore_unused(strategies);
+        return strategies.closest_points(point, box).apply(point, box);
+    }
+};
+*/
+
+template <closure_selector Closure>
+class point_to_range
+{
+public:
+
+    template <typename Point, typename Range, typename Segment, typename Strategies>
+    static inline void apply(Point const& point, Range const& range,
+                             Segment& shortest_seg,
+                             Strategies const& strategies)
+    {
+        using point_to_point_range = detail::closest_feature::point_to_point_range
+            <
+                Point, Range, Closure
+            >;
+
+        if (boost::size(range) == 0)
+        {
+            set_segment_from_points::apply(point, point, shortest_seg);
+            return;
+        }
+
+        closest_points::creturn_t<Point, Range, Strategies> cd_min;
+        
+        auto comparable_distance = strategy::distance::services::get_comparable
+            <
+                decltype(strategies.distance(point, range))
+            >::apply(strategies.distance(point, range));
+        
+        auto it_pair = point_to_point_range::apply(point, 
+                                                   boost::begin(range),
+                                                   boost::end(range), 
+                                                   comparable_distance,
+                                                   cd_min);
+        
+        auto closest_point = strategies.closest_points(point, range)
+            .apply(point, *it_pair.first, *it_pair.second);
```

Comment:
Minor, we usually use `it_pair->first` and `it_pair->second`

Reply:
Note that here `it_pair` is a pair of iterators not an iterator of a pair.

Reply:
If the name can be confusing maybe we could change it? This code is taken from `distance()` but here the type is not explicitly stated the only identifier that describes this variable in the code is its name. For me this either looks like an iterator to pair or a range represented as pair of iterators but is neither of these things. It's a representation of a segment. We already have names like `referring_segment` and `pointing_segment` in the code. My point is that the name should probably contain the word `segment`.

Reply:
ok I agree, what about `closest_segment` ? It is the closest segment in `range` to the query `point`.

Reply:
`closest_segment` looks good.